message/infrastructure/sql/schema.go

```diff
@@ -0,0 +1,211 @@
+	AckArgs(offset int, consumerGroup string) ([]interface{}, error)
+
+	// InsertQuery returns the SQL query that will insert the Watermill message into the SQL storage.
+	InsertQuery(topic string) string
+	// InsertArgs transforms the topic and Watermill message into the arguments put into InsertQuery.
+	InsertArgs(topic string, msg *message.Message) ([]interface{}, error)
+
+	// SelectQuery returns the SQL query that returns the next unread message for a given consumer group.
+	// Subscriber will not return those messages again for this consumer group.
+	SelectQuery(topic string) string
+	// SelectArgs transforms the topic into the argument put into SelectQuery.
+	SelectArgs(topic string, consumerGroup string) ([]interface{}, error)
+	// UnmarshalMessage transforms the Row obtained from the SQL query into a Watermill message.
+	// It also returns the offset of the last read message, for the purpose of acking.
+	UnmarshalMessage(row *sql.Row) (offset int, msg *message.Message, err error)
+
+	// EnsureTableForTopicQueries returns SQL query which will make sure (CREATE IF NOT EXISTS)
+	// that the tables exist to write messages to the given topic.
+	EnsureTableForTopicQueries(topic string) []string
+}
+
+// DefaultSchema is a default implementation of Inserter, Selecter and Acker that works with the following schema:
+//
+// `offset` bigint(20) NOT NULL AUTO_INCREMENT,
+// `uuid` binary(16) NOT NULL,
+// `created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+// `payload` json DEFAULT NULL,
+// `metadata` json DEFAULT NULL,
+// `topic` varchar(255) NOT NULL,
+type DefaultSchema struct {
+	Logger watermill.LoggerAdapter
+}
+
+func (s *DefaultSchema) EnsureTableForTopicQueries(topic string) []string {
+	messagesQ := strings.Join([]string{
+		"CREATE TABLE IF NOT EXISTS `watermill_" + topic + "` (",
+		"`offset` bigint(20) NOT NULL AUTO_INCREMENT PRIMARY KEY,",
+		"`uuid` binary(16) NOT NULL,",
+		"`created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,",
+		"`payload` json DEFAULT NULL,",
+		"`metadata` json DEFAULT NULL,",
+		"`topic` varchar(255) NOT NULL",
+		");",
+	}, "\n")
+
+	return []string{messagesQ}
+}
+
+func (s *DefaultSchema) InsertQuery(topic string) string {
+	if s.Logger == nil {
+		s.Logger = watermill.NopLogger{}
+	}
+
+	insertQ := strings.Join([]string{
+		`INSERT INTO`,
+		s.messagesTable(topic),
+		`(uuid, payload, metadata, topic) VALUES (?,?,?,?)`,
+	}, " ")
+
+	s.Logger.Info("Preparing query to insert messages", watermill.LogFields{
+		"q": insertQ,
+	})
+
+	return insertQ
+}
+
+func (s *DefaultSchema) InsertArgs(topic string, msg *message.Message) (args []interface{}, err error) {
+	logger := s.Logger
+	defer func() {
+		if err != nil {
+			logger.Error("Could not marshal message into SQL insert args", err, nil)
+			return
+		}
+		logger.Debug("Marshaled message into insert args", watermill.LogFields{
+			"uuid": msg.UUID,
+		})
+	}()
+
+	var uuid ulid.ULID
+	uuid, err = ulid.Parse(msg.UUID)
```

Comment:
what do you think to make it configurable? because it can be different to everyone (and not much people are using ulid)

Reply:
oooor we may store uuid as string?

Reply:
b29759fa0e4ea33dc1ad850755a9ddcfab9c343f no longer requiring ulid, just string